p2panda-rs/src/keypair.rs

```diff
@@ -73,16 +76,25 @@ impl KeyPair {
         Box::from(self.0.secret.to_bytes())
     }
 
-    /// Sign data using this key pair.
-    #[cfg(not(target_arch = "wasm32"))]
-    pub fn sign(&self, data: &[u8]) -> Signature {
-        self.0.sign(data)
+    /// Sign a message using this key pair.
+    pub fn sign(&self, message: &[u8]) -> Box<[u8]> {
+        Box::from(self.0.sign(message).to_bytes())
     }
 
     /// Verify a signature for a message.
     #[cfg(not(target_arch = "wasm32"))]
-    pub fn verify(&self, data: &[u8], signature: &Signature) -> Result<(), SignatureError> {
-        self.0.verify(data, signature)
+    pub fn verify(&self, message: &[u8], signature: &[u8]) -> Result<(), SignatureError> {
+        self.0
+            .verify(message, &Signature::try_from(signature).unwrap())
+    }
+
+    /// Verify a signature for a message.
+    #[cfg(target_arch = "wasm32")]
+    #[cfg_attr(target_arch = "wasm32", wasm_bindgen(js_name = verify))]
+    pub fn verify_js(&self, message: &[u8], signature: &[u8]) -> Result<(), JsValue> {
+        self.0
+            .verify(message, &Signature::try_from(signature).unwrap())
+            .map_err(|_| JsValue::from_str("invalid signature"))
```

Comment:
So I didn't find a way to make this work without having a separate function for the wasm target. `SignatureError` doesn't want to cross the wasm ABI. Can we implement some trait for `SignatureError` to make it compatible with wasm? Or use our own error type maybe? 

Also, failure to verify the signature from Javascript results in throwing a string instead of a proper `Error`. I didn't find a way to wrap the `JsValue` in `Err` or something like that and have it compile.

Reply:
Error handling in wasm for JavaScript seems to be a topic on itself, haven't read enough about it yet but I believe the wasm-bindgen documentation will have some words about it: https://rustwasm.github.io/docs/wasm-bindgen/reference/types/result.html?highlight=result#resultt-jsvalue

>  Can we implement some trait for SignatureError to make it compatible with wasm? Or use our own error type maybe?

For the library design in general I think having custom error types makes sense, it a) gives a clear API design of what to expect as a library consumer and b) gives us an understanding what we're dealing with developing it. So I'm in for using our own!

Reply:
But also, I think having two methods for two compiler targets is also totally fine and common. You could even name them the same :smile_cat: 

Btw, you can also write it like this as the first `cfg` attribute already sets the allowed target for the whole container:

```
/// Verify a signature for a message.
#[cfg(target_arch = "wasm32")]
#[wasm_bindgen(js_name = verify]
...
```

Reply:
Well, with two functions there's the possibility that they actually do different things. At least one of them should wrap the other then instead of replicating statements.

Regarding the error handling for javascript: throwing a string means that no traceback information is available and different errors can only be distinguished by pattern matching the error message. Not a huge problem per so but not elegant either. I would like to do [this](https://github.com/rustwasm/wasm-bindgen/issues/1742) but it doesn't seem to work for me.